### PR TITLE
*: avoid pass session ID by context & fix no schemaChecker error log

### LINKED ISSUE #22783: unexpected warning when inserting, schemaLeaseChecker is not set for this transaction
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

create table t(a bigint PRIMARY KEY auto_random);

cat 1.sql
INSERT INTO `t` VALUES ();

mysqlslap --query=1.sql -uroot -h127.0.0.1 -P4000 -c 100 --number-of-queries=10000000 --create-schema=test --commit=1

### 2. What did you expect to see? (Required)
No warning in tidb log.

### 3. What did you see instead (Required)
[2021/02/18 14:29:30.746 +08:00] [WARN] [2pc.go:1505] ["schemaLeaseChecker is not set for this transaction"] [conn=675] [sessionID=675] [startTS=423003362622177286] [commitTS=423003362622439424]
### 4. What is your TiDB version? (Required)
master

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve?

Issue Number: close #22783

Problem Summary:

Pass important argument by `context.WithValue` is a bad idea.

The callee need the data in the `context.Context`, but the caller maybe passing a `context.Bacground()`
#22783  is caused by such kind of refactoring.

In the beginning, we just use the session ID for logging.
Now I review the code and find we're using `sessionID == 0` to check whether it's an internal transaction or a user's transaction.

### What is changed and how it works?

Introduce a `txn.SetOption(kv.SessionID, id)` to replace the old way of passing session ID by `context.Context`.

What's Changed:

Use `txn.SetOption(kv.SessionID, id)` instead of `txn.Commit(tikvutil.SetSessionID(ctx, s.GetSessionVars().ConnectionID))` to pass the session ID.

How it Works:

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- No code

### Release note 

- No release note

### CODE REVIEW

File: store/tikv/txn.go [hunk 1]
```diff
@@ -369,9 +368,7 @@ func (txn *KVTxn) Commit(ctx context.Context) error {
 	// pessimistic transaction should also bypass latch.
 	if txn.store.txnLatches == nil || txn.IsPessimistic() {
 		err = committer.execute(ctx)
-		if val == nil || sessionID > 0 {
```
Comment (cfzjywxk) [inline store/tikv/txn.go:372]: Will this  make the following logic executed unexpected for the inner session statements whose session id is 0?

Comment (tiancaiamao) [reply store/tikv/txn.go:372]: For internal queries, `val` is always nil; for external queries `sessionID > 0` always hold,
so `val == nil || sessionID > 0` is basically always `true`? @MyonKeminta

Comment (tiancaiamao) [reply store/tikv/txn.go:372]: Now I've change it to:
 `sessionID > 0 ` => onCommit
`sessionID == 0` => nothing

The original author means to call `onCommit` for external queries (although `val == nil` may be confusing in the old code).
@cfzjywxk

File: executor/simple.go [hunk 2]
```diff
@@ -949,7 +948,8 @@ func (e *SimpleExec) executeAlterUser(s *ast.AlterUserStmt) error {
 		if err != nil {
 			return err
 		}
-		err = txn.Commit(tikvutil.SetSessionID(context.TODO(), e.ctx.GetSessionVars().ConnectionID))
+		txn.SetOption(kv.SessionID, e.ctx.GetSessionVars().ConnectionID)
+		err = txn.Commit(context.TODO())
```
Comment (lysu) [inline executor/simple.go:952]: does this make some log using `logutil.Logger(ctx)` miss connection id?
 
for example https://github.com/pingcap/tidb/blob/3ec8c8bae798642baf267601a5864238d3367b9e/store/tikv/2pc.go#L1033

File: executor/adapter.go [hunk 3]
```diff
@@ -306,7 +306,6 @@ func (a *ExecStmt) Exec(ctx context.Context) (_ sqlexec.RecordSet, err error) {
 	}()
 
 	sctx := a.Ctx
-	ctx = util.SetSessionID(ctx, sctx.GetSessionVars().ConnectionID)
```
Comment (lysu) [inline executor/adapter.go:309]: it seems some other test https://github.com/pingcap/tidb/blob/18cbfaac15f8478902726e1e64971aa96b862462/store/tikv/region_request.go#L413 need this ..

@MyonKeminta  PTAL

File: store/tikv/txn.go (hunk 1 above)
Comment (lysu) [reply store/tikv/txn.go:372]: maybe it's hard to distinguish external from internal queries form `sessionID`?

InRestrictedSQL also can call doCommit

https://github.com/pingcap/tidb/blob/c3f9b0899d2a5c7ab8da0e37f358596f2fe27913/session/session.go#L521

Comment (tiancaiamao) [reply store/tikv/txn.go:372]: If we take a session from the session pool to execute the SQL, `sessionID` is not set.
However, if we direct use the session to execute a internal SQL, that might be a problem.

File: executor/adapter.go (hunk 3 above)
Comment (tiancaiamao) [reply executor/adapter.go:309]: I have updated the test, you can see the `_test.go` file changes in this PR

Comment (lysu) [reply executor/adapter.go:309]: it seems the previous 
https://github.com/pingcap/tidb/blob/18cbfaac15f8478902726e1e64971aa96b862462/store/tikv/region_request.go#L465
will be used in some test case that out of tidb repo 
https://github.com/pingcap/automated-tests/blob/cc44a56fd59cfdb71eaf37210e0861ba7acdc925/ticases/transaction/asynccommit/linearizability.go#L445

I'm not sure whether it works  😞

Comment (cfzjywxk) [reply executor/adapter.go:309]: Seems we may change the session id or abstract a `isInnerExecution` accordingly.

Comment (MyonKeminta) [reply executor/adapter.go:309]: Yes that's a problem.  Some failpoints that's used by tests outside the repo checks the session id and works only when session id != 0. These kind of usages also occurs in 2pc.go, prewrite.go, etc.

**Comment by tiancaiamao** [comment]
Close for now, it's not a very easy fix.

**Pull request closed without merging**